### PR TITLE
NGDOC-134 added introduction to LB7 and ES

### PR DESCRIPTION
**Pull request by nibinqtl**
And some other related changes on other pages.

### CODE REVIEW

File: en/docs/edge-logic/built-in-variables.md [hunk 1]
```diff
@@ -39,14 +39,15 @@ If you use the CDN Pro portal to edit the property, the autocompletion feature w
...
 | <span id="request"></span><span class="var">$request</span> | full HTTP request line | <span class="badge green">Yes</span> | <span class="badge green">Yes</span> |
-| <span id="request_cpu_time"></span><span class="var">$request_cpu_time</span> |<span class="badge small" title="numerical value">#</span> CPU time, in nanoseconds, spent on processing the request by the CS stage (when used in the Edge Logic), or by the LB stage (when used in the real-time log) (ETA: Oct. 2021)| <span class="badge green">Yes</span> | <span class="badge green">Yes</span> |
+| <span id="request_cpu_time"></span><span class="var">$request_cpu_time</span> |<span class="badge small" title="numerical value">#</span> CPU time, in nanoseconds, spent on processing the request by the CS stage (when used in the Edge Logic), or by the LB stage (when used in the real-time log) | <span class="badge green">Yes</span> | <span class="badge green">Yes</span> |
```
Comment (burtonquantil) [inline en/docs/edge-logic/built-in-variables.md:50]: The commas around nanoseconds are probably not really necessary though it's OK.

@marcfishelquantil ?

File: en/docs/edge-logic/built-in-variables.md [hunk 2]
```diff
@@ -39,14 +39,15 @@ If you use the CDN Pro portal to edit the property, the autocompletion feature w
 | <span id="pid"></span><span class="var">$pid</span> | process ID of the service | <span class="badge green">Yes</span> | <span class="badge green">Yes</span> |
 | <span id="pipe"></span><span class="var">$pipe</span> | "p" if request is pipelined, or "." otherwise. | <span class="badge yellow">No</span> | <span class="badge green">Yes</span> |
 | <span id="property_ver"></span><span class="var">$property_ver</span> |<span class="badge small" title="numerical value">#</span> the property version number | <span class="badge green">Yes</span> | <span class="badge green">Yes</span> |
+| <span id="qtl_cpu_cycle"></span><span class="var">$qtl_cpu_cycle</span> |<span class="badge small" title="numerical value">#</span> CPU Time in nanoseconds（1e-9s) spent in LB7, and in ES before transferring the response body| <span class="badge yellow">No</span> | <span class="badge green">Yes</span> |
```
Comment (burtonquantil) [inline en/docs/edge-logic/built-in-variables.md:42]: "Time" should be "time"

File: en/docs/edge-logic/declarative-imperative.md [hunk 3]
```diff
@@ -0,0 +1 @@
+## Declarative and Imperative Directives
```
Comment (burtonquantil) [inline en/docs/edge-logic/declarative-imperative.md:1]: Are you adding anything else to this file?

File: en/docs/edge-logic/lb7-es-structure.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+## The Layer 7 Load Balancer - Edge Server Structure
+
+The edge network of CDN Pro consists of hundreds of nodes around the world and in each
```
Comment (burtonquantil) [inline en/docs/edge-logic/lb7-es-structure.md:3]: How about this rewrite: "CDN Pro's edge network consists of hundreds of nodes around the world, each of which is a farm of physical server."

File: en/docs/edge-logic/lb7-es-structure.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+node is a farm of physical servers. To most efficiently utilize these servers, we need our design to meet the following criteria:
```
Comment (burtonquantil) [inline en/docs/edge-logic/lb7-es-structure.md:4]: Suggestion:  "To most efficiently utilize these servers, our design satisfies the following criteria:"

File: en/docs/edge-logic/lb7-es-structure.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+1. The service load needs to be distributed as evenly as possible across all servers. The 
+'load' includes network bandwidth, CPU power and disk usage.
```
Comment (burtonquantil) [inline en/docs/edge-logic/lb7-es-structure.md:7]: Minor:  I'd add a comma after "power"

**Review by burtonquantil — CHANGES_REQUESTED** [review]
Please check my suggestions.

File: en/docs/edge-logic/built-in-variables.md [hunk 7]
```diff
@@ -24,7 +24,7 @@ If you use the CDN Pro portal to edit the property, the autocompletion feature w
 | <span id="dollar_sign"></span><span class="var">$dollar_sign</span> | a literal dollar sign | <span class="badge green">Yes</span> | <span class="badge green">Yes</span> |
 | <span id="download_time"></span><span class="var">$download_time</span> |<span class="badge small" title="numerical value">#</span> the time from CDN Pro receives the client request to CDN Pro sends the response body to the system buffer, the value should be equal to sum of $request_end_time/$turn_around_time/$transfer_time, in seconds. Another variable download_time_ms shares the same meaning but in milliseconds.  | <span class="badge yellow">No</span> | <span class="badge green">Yes</span>
```
Comment (marcfishelquantil) [inline en/docs/edge-logic/built-in-variables.md:25]: May I suggest replacing "...the time from CDN Pro receives the client request to CDN Pro..." with "...the time from when CDN Pro receives the client request until CDN Pro..."

File: en/docs/edge-logic/built-in-variables.md [hunk 8]
```diff
@@ -1,9 +1,9 @@
 ## Built-in Variables
 
-The table below lists all the built-in variables supported by the CDN Pro servers. You can use them in Edge Logic or the real-time log. Not all variables are supported in both places. The variables with numerical values are tagged with <span class="badge small" title="numerical value">#</span>.
+The table below lists all the built-in variables supported by the CDN Pro servers. You can use them in Edge Logic or [Load Balancer Logic](lb7-es-structure). Not all variables are supported in both places. The [real-time log](/docs/portal/edge-configurations/creating-property#real-time-log) supports the same set of variables as the Load Balancer Logic. The variables with numerical values are tagged with <span class="badge small" title="numerical value">#</span>.
 If you use the CDN Pro portal to edit the property, the autocompletion feature will list all the supported variables if you type `$` at a position where a variable is allowed.
```
Comment (marcfishelquantil) [inline en/docs/edge-logic/built-in-variables.md:4]: How about rewording this to: "If you use the CDN Pro portal to edit the property, typing `$` at a position where a variable is allowed will list all the supported variables"

File: en/docs/edge-logic/built-in-variables.md (hunk 1 above)
Comment (marcfishelquantil) [reply en/docs/edge-logic/built-in-variables.md:50]: @burtonquantil Please let's retain the commas.

File: en/docs/edge-logic/lb7-es-structure.md (hunk 4 above)
Comment (marcfishelquantil) [reply en/docs/edge-logic/lb7-es-structure.md:3]: @burtonquantil In your suggested rewrite, should "physical server" be plural (physical servers)?

**Review by marcfishelquantil — COMMENTED** [review]
Bin, I hope my comments help.

File: en/docs/edge-logic/lb7-es-structure.md (hunk 4 above)
Comment (burtonquantil) [reply en/docs/edge-logic/lb7-es-structure.md:3]: @marcfishelquantil  yes..

File: en/docs/edge-logic/declarative-imperative.md (hunk 3 above)
Comment (nibinqtl) [reply en/docs/edge-logic/declarative-imperative.md:1]: This is a placeholder where I plan to add content in the future

**Comment by nibinqtl** [comment]
Thanks. just addressed all your comments.

File: en/docs/edge-logic/lb7-es-structure.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+2. For cacheable objects, avoid caching duplicated copies in the same farm.
```
[the remark this answers is not in the document]
Comment (burtonquantil) [reply en/docs/edge-logic/lb7-es-structure.md:7]: @nibinqtl any reason why this shouldn't be "duplicate"?

File: en/docs/edge-logic/lb7-es-structure.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+We end up with the following logical structure in each edge node:
+<p align=center src=“https://docs.google.com/drawings/d/1Q-UdhK7w5Yzkjg53NTlzYAzKOOBmW_Rou5IgdGvl7-s/edit”><img src="/docs/edge-logic/lb7-es-structure.png" alt="Node Logical Structure" width="600"></p>
```
[the remark this answers is not in the document]
Comment (burtonquantil) [reply en/docs/edge-logic/lb7-es-structure.md:10]: @nibinqtl  Any reason why we point to your Google doc? It seems unnecessary.

File: en/docs/edge-logic/lb7-es-structure.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+Each request admitted by the router first goes through the layer 4 firewall and load balancer and reaches
+a layer 7 load balancer (LB7). The LB7 terminates the TLS tunnel, performs some initial processing of the
+HTTP request, then distribute it to an edge server (ES) for further processing. The distribution is
+based on a consistent hash algorithm and the hash key involves the request URI and a [customizable](/docs/portal/edge-configurations/creating-property#advanced-settings)
```
[the remark this answers is not in the document]
Comment (burtonquantil) [reply en/docs/edge-logic/lb7-es-structure.md:15]: @nibinqtl suggested change not accepted?

File: en/docs/edge-logic/lb7-es-structure.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+variable to make sure the ES are evenly loaded. ES does most of the processing of the requests including
+everything we put in the edge logic. It also launches requests to contact the origin, parent or 
+external servers to obtain necessary data. Any cacheable objects are stored in the ES.
+
+Based on this structure, we can see that some tasks will be more efficiently performed in
+the LB7 than in the ES. If a request can be responded in
```
[the remark this answers is not in the document]
Comment (burtonquantil) [reply en/docs/edge-logic/lb7-es-structure.md:21]: @nibinqtl  how about the above suggestion?s

**Review by burtonquantil — CHANGES_REQUESTED** [review]
You didn't respond to a few of my earlier suggestions.
I didn't see any comment from Marc saying they were incorrect suggestions.

File: en/docs/edge-logic/lb7-es-structure.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (nibinqtl) [reply en/docs/edge-logic/lb7-es-structure.md:10]: this reminds people (including myself) where to find the source of the image in case we need to change it in the future.

**Comment by nibinqtl** [comment]
It looks not all comments are displayed on the "Conversation" tab...

**Review by burtonquantil — APPROVED** [review]
OK to merge
